File: Google Code Jam/2013/Qualification Round/C.cc
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <algorithm>

using namespace std;

vector<long long> valid;

bool check(long long x)
{
    ostringstream os;
    os << x;
    string s = os.str();
    for (int i = 0, j = (int)s.length() - 1; i < j; i++, j--) {
        if (s[i] != s[j]) return false;
    }
    return true;
}

int count(long long x)
{
    return upper_bound(valid.begin(), valid.end(), x) - valid.begin();
}

int main()
{
    const int UPBOUND = 1e7 + 10;
    for (int i = 0; i < UPBOUND; i++) {
        if (check(i) && check((long long)i * i)) {
            valid.push_back((long long)i * i);
            // cout << (long long)i * i << endl;
        }
    }

    int tasks;
    cin >> tasks;
    for (int task = 1; task <= tasks; task++) {
        long long A, B;
        cin >> A >> B;
        cout << "Case #" << task << ": " << count(B) - count(A - 1) << endl;
    }

    return 0;
}

File: Google Code Jam/2013/Qualification Round/C.rb
def check(x)
  x = x.to_s
  x == x.reverse
end

def count(x)
  l, r = 0, @valid.length
  while l < r
    mid = (l + r - 1) >> 1
    if @valid[mid] <= x
      l = mid + 1
    else
      r = mid
    end
  end
  l
end

@valid = []
10000010.times { |x| @valid << x * x if check(x) && check(x * x) }

tasks = gets.to_i
tasks.times do |task|
  a, b = gets.split(" ").map(&:to_i)
  puts "Case ##{task + 1}: #{count(b) - count(a - 1)}"
end
